File: backend/src/Renderer.cpp
// renderer.cpp

#include "Renderer.h"
#include "Error.h"
#include "CmdSave.h"

#include <QDebug>
#include <QImage>
#include <QPdfWriter>
#include <QSvgGenerator>
#include <QImage>
#include <cmath>
#include <QFileInfo>
#include <QDir>
#include <QTemporaryFile>

Renderer::Renderer() {
  maxtries = 1000;
  bitmapres = 300;
  bitmapqual = 95;
  overridewidth = 0;
  overrideheight = 0;
}

void Renderer::prerender(int upto) {
  //  qDebug() << "prerender" << upto;
  QImage img(1,1,QImage::Format_ARGB32);
  fig.setSize(QSizeF(1, 1)); // this may be overridden later
  fig.painter().begin(&img);
  fig.hardReset();
  foreach (QString p, prog.panels(upto)) {
    QRectF dataExtent = prog.dataRange(p, upto);
    if (p=="-") {
      fig.xAxis().setDataRange(dataExtent.left(), dataExtent.right());
      fig.yAxis().setDataRange(dataExtent.top(), dataExtent.bottom());
    } else {
      fig.panelRef(p).xaxis.setDataRange(dataExtent.left(),
                                         dataExtent.right());
      fig.panelRef(p).yaxis.setDataRange(dataExtent.top(),
                                         dataExtent.bottom());
    }
  }

  QMap<int, int> itercount; // organize by line number
  QMap<int, int> maxitercount; // organize by line number
  bool fail = false;
  int totaliter = 0;
  while (true) {
    int line = prog.render(fig, true, upto);
    // renderer to determine paper bbox & fudge
    if (fig.checkFudgeFailure()) {
      fail = true;
      break;
    } else if (fig.checkFudged()) {
      for (auto it=itercount.begin(); it!=itercount.end(); ++it)
        if (it.key()<line)
          it.value() = 0;
      itercount[line] += 1;
      totaliter += 1;
      if (itercount[line] > maxitercount[line])
        maxitercount[line] = itercount[line];
      if (itercount[line] >= maxtries) {
        Error() << QString("Shrink iterations exceeded at %1: %2 >= %3")
          .arg(line).arg(itercount[line]).arg(maxtries);
        fail = true;
        break;
      }
    } else {
      break;
    }
  } 


  if (fail) {
    Error() << QString("\"Shrink\" failed after %1 iterations").arg(totaliter);
    qDebug() << "Iteration detail: " << itercount;
    qDebug() << "Max iteration detail: " << maxitercount;
  }

  fig.painter().end();
}


bool Renderer::renderSVG(QString ofn, int upto) {
  QSvgGenerator img;
  img.setFileName(ofn);
  img.setResolution(90); // anything else seems to be poorly supported
  img.setViewBox(QRectF(QPointF(0,0),
			QSizeF(90./72*iu2pt(fig.extent().width()),
			       90./72*iu2pt(fig.extent().height()))));
  if (!fig.painter().begin(&img))
    return false;
  fig.setHairline(0.25);
  fig.painter().scale(90./72*iu2pt(), 90./72*iu2pt());
  fig.setDashScale(90./72*iu2pt());
  fig.painter().translate(fig.extent().left(),
			  fig.extent().top());
  prog.render(fig, false, upto);
  fig.painter().end();
  return true;
}


bool Renderer::renderPDF(QString ofn, int upto) {
  QPdfWriter img(ofn);
  img.setPageSize(QPageSize(QSizeF(iu2pt(fig.extent().width()),
				            iu2pt(fig.extent().height())), 
							QPageSize::Point));

  if (!fig.painter().begin(&img))
    return false;
  fig.setHairline(0.25);

  double dpix = img.logicalDpiX();
  double dpiy = img.logicalDpiY();

  fig.painter().translate(-10*dpix/72., -10*dpiy/72.);
  // I don't know why this translation is needed.

  fig.painter().scale(iu2pt()*dpix/72.0,
		      iu2pt()*dpix/72.0);
  fig.setDashScale(iu2pt()*std::sqrt(dpix*dpiy)/72.0);
  fig.painter().translate(-fig.extent().left(),
			  -fig.extent().top());
  prog.render(fig, false, upto);
  fig.painter().end();
  return true;
}



bool Renderer::renderImage(QString ofn, int upto) {
  QImage img(int(fig.extent().width()),
	     int(fig.extent().height()),
	     QImage::Format_ARGB32);
  img.fill(0xffffffff);
  fig.painter().begin(&img);
  fig.painter().setRenderHint(QPainter::Antialiasing);
  fig.painter().setRenderHint(QPainter::TextAntialiasing);
  fig.setHairline(0);

  fig.setDashScale(1);
  fig.painter().translate(-fig.extent().left(),
			  -fig.extent().top());
  prog.render(fig, false, upto); 
  fig.painter().end();
  if (ofn.endsWith("jpg") || ofn.endsWith("jpeg"))
    return img.save(ofn, 0, bitmapqual);
  else
    return img.save(ofn);
}


static bool error(QString const &s) {
  Error() << s;
  return false;
}

static bool error_failtosave(QString const &fn) {
  if (fn.isEmpty())
    return error("Failed to save: no filename");
  QFileInfo fi(fn);
  QDir d(fi.dir());
  QString err = QString("Failed to save as “%1”").arg(fn);
  if (!d.exists())
    err += QString(": The folder “%1” does not exist.").arg(d.path());
  else if (!QFileInfo(d.path()).isWritable())
    err += QString(": The folder “%1” is not writable.").arg(d.path());
  else if (fi.exists() && !fi.isWritable())
    err += ": File exists and is not writable.";
  else
    err += ". (Reason unknown.)";
  return error(err);
}

static bool error_unknownextension(QString const &extn) {
  QString err = "Failed to save";
  if (extn.isEmpty())
    err += ": Filename without an extension.";
  else
    err += QString(": Unknown extension “%1”.").arg(extn);
  return error(err);
}


bool Renderer::save(QString ofn, int upto) {
  if (ofn=="-")
    ofn = "-.pdf";
  
  int idx = ofn.lastIndexOf(".");
  if (idx<0)
    return error("Output file must have an extension");
  QString extn = ofn.mid(idx+1);

  bool extnIsBitmap = extn=="png" || extn=="jpg"
    || extn=="tif" || extn=="tiff";
  if (extnIsBitmap)
    setFACTOR(bitmapres/72);

  QTemporaryFile tmpf;
  bool usetmpf = false;
  if (ofn == "-." + extn) {
    // write to stdout
    if (!tmpf.open())
      return error("Cannot write to temporary file");
    tmpf.close();
    ofn = tmpf.fileName();
    usetmpf = true;
  }

  Figure fig;
  fig.overrideSize(QSizeF(pt2iu(overridewidth),
			  pt2iu(overrideheight)));
  prerender(upto);

  if (extn == "svg") {
    if (!renderSVG(ofn, upto))
      return error_failtosave(ofn);
  } else if (extn == "pdf") {
    if (!renderPDF(ofn, upto))
      return error_failtosave(ofn);
  } else if (extnIsBitmap) {
    if (!renderImage(ofn, upto))
      return error_failtosave(ofn);
  } else {
    if (extn=="eps" || extn=="ps") 
      return error("PostScript output is no longer supported");
    else
      return error_unknownextension(extn);
  }

  if (usetmpf) {
    if (!tmpf.open())
      error("Cannot re-read temporary file");
    QFile f;
    if (!f.open(stdout, QFile::WriteOnly))
      error("Cannot write to stdout");
    while (1) {
      QByteArray ba = tmpf.read(1024*1024);
      if (ba.isEmpty())
	break;
      if (f.write(ba)!=ba.size())
	error("Failure to write to stdout");
    }
    if (!tmpf.atEnd())
      error("Error reading from temporary file");
  }
  return true;
}  


void Renderer::setMaxTries(int n) {
  //  qDebug() << "setmaxtries" << n;
  maxtries = n;
}

void Renderer::overrideWidth(double w) {
  overridewidth = w;
}

void Renderer::overrideHeight(double h) {
  overrideheight =h;
}

void Renderer::setBitmapResolution(double r) {
  bitmapres = r;
}

void Renderer::setBitmapQuality(int q) {
  bitmapqual = q;
}

//void Renderer::perhapsSave() {
//  CmdSave *cmd = prog.nextSave();
//  if (cmd) {
//    setBitmapResolution(cmd->resolution());
//    setBitmapQuality(cmd->quality());
//    save(cmd->filename());
//  }
//}

void Renderer::dosaves(int start, int end) {
  if (end<0)
    end = prog.length();
  for (int k=start; k<end; k++) {
    CmdSave const *cmd = dynamic_cast<CmdSave const *>(prog.command(k));
    if (cmd) {
      setBitmapResolution(cmd->resolution());
      setBitmapQuality(cmd->quality());
      save(cmd->filename(), k);
    }
  }
}

File: backend/python/renderer.py
# renderer.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

import math
import os
import sys
import tempfile

from PyQt6.QtCore import QPointF, QRectF, QSizeF
from PyQt6.QtGui import QImage, QPainter, QPageSize
from PyQt6.QtSvg import QSvgGenerator
from PyQt6.QtPdf import QPdfWriter

from .program import Programfrom .figure import Figurefrom .factor import pt2iu, iu2pt, set_factorfrom .error import Errorfrom .cmdsave import CmdSave

class Renderer:
    """Orchestrates pre-rendering (fudge loop) and output to SVG/PDF/image."""

    def __init__(self) -> None:
        self._prog:            Program = Program()
        self._fig:             Figure  = Figure()
        self._maxtries:        int     = 1000
        self._bitmapres:       float   = 300.0
        self._bitmapqual:      int     = 95
        self._overridewidth:   float   = 0.0
        self._overrideheight:  float   = 0.0

    # --- Accessors --------------------------------------------------------

    def program(self) -> Program:
        return self._prog

    def figure(self) -> Figure:
        return self._fig

    # --- Pre-render (fudge/shrink iteration loop) -------------------------

    def prerender(self, upto: int = -1) -> None:
        img = QImage(1, 1, QImage.Format.Format_ARGB32)
        self._fig.set_size(QSizeF(1, 1))
        self._fig.painter().begin(img)
        self._fig.hard_reset()

        for p in self._prog.panels(upto):
            data_extent = self._prog.data_range(p, upto)
            if p == "-":
                self._fig.x_axis().set_data_range(
                    data_extent.left(), data_extent.right())
                self._fig.y_axis().set_data_range(
                    data_extent.top(), data_extent.bottom())
            else:
                self._fig.panel_ref(p).xaxis.set_data_range(
                    data_extent.left(), data_extent.right())
                self._fig.panel_ref(p).yaxis.set_data_range(
                    data_extent.top(), data_extent.bottom())

        iter_count:     dict[int, int] = {}
        max_iter_count: dict[int, int] = {}
        fail       = False
        total_iter = 0

        while True:
            line = self._prog.render(self._fig, dryrun=True, upto=upto)
            if self._fig.check_fudge_failure():
                fail = True
                break
            elif self._fig.check_fudged():
                for k in list(iter_count):
                    if k < line:
                        iter_count[k] = 0
                iter_count[line] = iter_count.get(line, 0) + 1
                total_iter += 1
                max_iter_count[line] = max(
                    max_iter_count.get(line, 0), iter_count[line])
                if iter_count[line] >= self._maxtries:
                    Error() << (f"Shrink iterations exceeded at {line}: "
                                f"{iter_count[line]} >= {self._maxtries}")
                    fail = True
                    break
            else:
                break

        if fail:
            Error() << f'"Shrink" failed after {total_iter} iterations'
            print(f"Iteration detail: {iter_count}", file=sys.stderr)
            print(f"Max iteration detail: {max_iter_count}", file=sys.stderr)

        self._fig.painter().end()

    # --- Format-specific renderers ----------------------------------------

    def render_svg(self, ofn: str, upto: int = -1) -> bool:
        img = QSvgGenerator()
        img.setFileName(ofn)
        img.setResolution(90)
        img.setViewBox(QRectF(
            QPointF(0, 0),
            QSizeF(90. / 72 * iu2pt(self._fig.extent().width()),
                   90. / 72 * iu2pt(self._fig.extent().height())),
        ))
        if not self._fig.painter().begin(img):
            return False
        self._fig.set_hairline(0.25)
        self._fig.painter().scale(90. / 72 * iu2pt(),
                                  90. / 72 * iu2pt())
        self._fig.set_dash_scale(90. / 72 * iu2pt())
        self._fig.painter().translate(self._fig.extent().left(),
                                      self._fig.extent().top())
        self._prog.render(self._fig, dryrun=False, upto=upto)
        self._fig.painter().end()
        return True

    def render_pdf(self, ofn: str, upto: int = -1) -> bool:
        img = QPdfWriter(ofn)
        img.setPageSize(QPageSize(
            QSizeF(iu2pt(self._fig.extent().width()),
                   iu2pt(self._fig.extent().height())),
            QPageSize.Unit.Point,
        ))
        if not self._fig.painter().begin(img):
            return False
        self._fig.set_hairline(0.25)
        dpix = img.logicalDpiX()
        dpiy = img.logicalDpiY()
        self._fig.painter().translate(-10 * dpix / 72., -10 * dpiy / 72.)
        self._fig.painter().scale(iu2pt() * dpix / 72.0,
                                  iu2pt() * dpix / 72.0)
        self._fig.set_dash_scale(iu2pt() * math.sqrt(dpix * dpiy) / 72.0)
        self._fig.painter().translate(-self._fig.extent().left(),
                                      -self._fig.extent().top())
        self._prog.render(self._fig, dryrun=False, upto=upto)
        self._fig.painter().end()
        return True

    def render_image(self, ofn: str, upto: int = -1) -> bool:
        img = QImage(
            int(self._fig.extent().width()),
            int(self._fig.extent().height()),
            QImage.Format.Format_ARGB32,
        )
        img.fill(0xFFFFFFFF)
        self._fig.painter().begin(img)
        self._fig.painter().setRenderHint(QPainter.RenderHint.Antialiasing)
        self._fig.painter().setRenderHint(QPainter.RenderHint.TextAntialiasing)
        self._fig.set_hairline(0)
        self._fig.set_dash_scale(1)
        self._fig.painter().translate(-self._fig.extent().left(),
                                      -self._fig.extent().top())
        self._prog.render(self._fig, dryrun=False, upto=upto)
        self._fig.painter().end()
        if ofn.endswith("jpg") or ofn.endswith("jpeg"):
            return img.save(ofn, quality=self._bitmapqual)
        return img.save(ofn)

    # --- Top-level save ---------------------------------------------------

    def save(self, ofn: str, upto: int = -1) -> bool:
        if ofn == "-":
            ofn = "-.pdf"

        idx = ofn.rfind(".")
        if idx < 0:
            return _error("Output file must have an extension")
        extn = ofn[idx + 1:]

        extn_is_bitmap = extn in ("png", "jpg", "tif", "tiff")
        if extn_is_bitmap:
            set_factor(self._bitmapres / 72)

        # Handle stdout output via a temporary file
        use_tmpf = False
        tmpf_path = ""
        if ofn == f"-.{extn}":
            fd, tmpf_path = tempfile.mkstemp(suffix=f".{extn}")
            os.close(fd)
            ofn = tmpf_path
            use_tmpf = True

        self._fig = Figure()
        self._fig.override_size(QSizeF(pt2iu(self._overridewidth),
                                       pt2iu(self._overrideheight)))
        self.prerender(upto)

        ok = False
        if extn == "svg":
            ok = self.render_svg(ofn, upto)
        elif extn == "pdf":
            ok = self.render_pdf(ofn, upto)
        elif extn_is_bitmap:
            ok = self.render_image(ofn, upto)
        elif extn in ("eps", "ps"):
            return _error("PostScript output is no longer supported")
        else:
            return _error_unknown_extension(extn)

        if not ok:
            return _error_fail_to_save(ofn)

        if use_tmpf:
            try:
                with open(tmpf_path, "rb") as f:
                    while True:
                        chunk = f.read(1024 * 1024)
                        if not chunk:
                            break
                        sys.stdout.buffer.write(chunk)
                sys.stdout.buffer.flush()
            except OSError as e:
                _error(f"Failed to write to stdout: {e}")
            finally:
                os.unlink(tmpf_path)

        return True

    # --- Iteration and override controls ----------------------------------

    def set_max_tries(self, n: int) -> None:
        self._maxtries = n

    def override_width(self, w: float) -> None:
        self._overridewidth = w

    def override_height(self, h: float) -> None:
        self._overrideheight = h

    def set_bitmap_resolution(self, r: float) -> None:
        self._bitmapres = r

    def set_bitmap_quality(self, q: int) -> None:
        self._bitmapqual = q

    def dosaves(self, start: int = 0, end: int = -1) -> None:
        """Execute all CmdSave commands in the given range."""
        limit = len(self._prog) if end < 0 else end
        for k in range(start, limit):
            cmd = self._prog.command(k)
            if isinstance(cmd, CmdSave):
                self.set_bitmap_resolution(cmd.resolution())
                self.set_bitmap_quality(cmd.quality())
                self.save(cmd.filename(), k)


# ---------------------------------------------------------------------------
# Module-level error helpers
# ---------------------------------------------------------------------------

def _error(s: str) -> bool:
    Error() << s
    return False


def _error_fail_to_save(fn: str) -> bool:
    if not fn:
        return _error("Failed to save: no filename")
    err = f'Failed to save as \u201c{fn}\u201d'
    d = os.path.dirname(os.path.abspath(fn))
    if not os.path.isdir(d):
        err += f': The folder \u201c{d}\u201d does not exist.'
    elif not os.access(d, os.W_OK):
        err += f': The folder \u201c{d}\u201d is not writable.'
    elif os.path.exists(fn) and not os.access(fn, os.W_OK):
        err += ': File exists and is not writable.'
    else:
        err += '. (Reason unknown.)'
    return _error(err)


def _error_unknown_extension(extn: str) -> bool:
    if not extn:
        return _error('Failed to save: Filename without an extension.')
    return _error(f'Failed to save: Unknown extension \u201c{extn}\u201d.')
